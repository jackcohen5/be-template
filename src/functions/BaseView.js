import { errorResponse, successResponse } from 'services/Lambda'

const BaseView = f => async (event, ...otherParams) => {
    try {
        const enhancedEvent = {
            itemId: event.pathParameters ? event.pathParameters.itemId : null,
            userId:
                event.requestContext && event.requestContext.authorizer
                    ? event.requestContext.authorizer.userId
                    : null,
        }
        const { data, statusCode = 200 } = await f(
            enhancedEvent,
            ...otherParams,
        )
        return successResponse(
            {
                data,
            },
            statusCode,
        )
    } catch (err) {
        console.error(err)
        return errorResponse({
            message: err.message,
        })
    }
}

export default BaseView
