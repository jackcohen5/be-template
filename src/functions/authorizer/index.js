import { promisify } from 'util'

import jwt from 'jsonwebtoken'
import jwksClient from 'jwks-rsa'

import {
    AUTH0_AUDIENCE,
    AUTH0_DOMAIN,
    ROLES_CLAIM_KEY,
} from './authorizer.constants'

const getSigningKey = async kid => {
    const client = jwksClient({
        jwksUri: `${AUTH0_DOMAIN}.well-known/jwks.json`,
    })
    return promisify(client.getSigningKey)(kid)
}

const generatePolicy = ({ effect, resource, roles = [], userId }) => {
    const authResponse = {
        principalId: userId,
        context: {
            userId,
            roles,
        },
    }

    if (effect && resource) {
        authResponse.policyDocument = {
            Version: '2012-10-17',
            Statement: [
                {
                    Action: 'execute-api:Invoke',
                    Effect: effect,
                    Resource: resource,
                },
            ],
        }
    }

    return authResponse
}

const Authorize = async event => {
    try {
        const splitAuthHeader = event.authorizationToken.split(' ')
        const token =
            splitAuthHeader.length > 1
                ? splitAuthHeader[1]
                : event.authorizationToken
        const decoded = jwt.decode(token, { complete: true })
        try {
            if (decoded.payload.iss !== AUTH0_DOMAIN) {
                throw new Error('Auth Error: JWT issuer does not match domain')
            }

            const signingKey = await getSigningKey(decoded.header.kid)
            const key = signingKey.publicKey || signingKey.rsaPublicKey
            jwt.verify(token, key, {
                algorithm: ['RS256'],
                audience: AUTH0_AUDIENCE,
            })
            return generatePolicy({
                effect: 'Allow',
                resource: '*',
                roles: decoded.payload[ROLES_CLAIM_KEY],
                userId: decoded.payload.sub,
            })
        } catch (err) {
            console.error(err)
            return generatePolicy({
                effect: 'Deny',
                resource: event.methodArn,
                userId: decoded.payload.sub,
            })
        }
    } catch (err) {
        console.error(err)
        return generatePolicy({
            effect: 'Deny',
            resource: event.methodArn,
            userId: 'Unauthorized',
        })
    }
}

export default Authorize
