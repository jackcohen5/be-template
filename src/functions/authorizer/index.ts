import {
    APIGatewayAuthorizerResult,
    APIGatewayTokenAuthorizerHandler,
} from 'aws-lambda'
import jwt from 'jsonwebtoken'
import * as jwksClient from 'jwks-rsa'

import {
    AUTH0_AUDIENCE,
    AUTH0_DOMAIN,
    ROLES_CLAIM_KEY,
} from './authorizer.constants'

const generatePolicy = ({
    effect,
    resource,
    roles = [],
    userId,
}): APIGatewayAuthorizerResult => {
    return {
        context: {
            userId,
            roles: JSON.stringify(roles),
        },
        policyDocument: {
            Version: '2012-10-17',
            Statement: [
                {
                    Action: 'execute-api:Invoke',
                    Effect: effect,
                    Resource: resource,
                },
            ],
        },
        principalId: userId,
    }
}

const getPublicKey = async (kid) => {
    const client = jwksClient({
        jwksUri: `${AUTH0_DOMAIN}.well-known/jwks.json`,
    })
    const signingKey = await client.getSigningKey(kid)
    return signingKey.getPublicKey()
}

const Authorize: APIGatewayTokenAuthorizerHandler = async (event) => {
    try {
        const splitAuthHeader = event.authorizationToken.split(' ')
        const token =
            splitAuthHeader.length > 1
                ? splitAuthHeader[1]
                : event.authorizationToken
        const decoded = jwt.decode(token, { complete: true })
        try {
            if (decoded.payload.iss !== AUTH0_DOMAIN) {
                throw new Error('Auth Error: JWT issuer does not match domain')
            }

            const key = await getPublicKey(decoded.header.kid)
            jwt.verify(token, key, {
                algorithm: ['RS256'],
                audience: AUTH0_AUDIENCE,
            })
            return generatePolicy({
                effect: 'Allow',
                resource: '*',
                roles: decoded.payload[ROLES_CLAIM_KEY],
                userId: decoded.payload.sub,
            })
        } catch (err) {
            console.error(err)
            return generatePolicy({
                effect: 'Deny',
                resource: event.methodArn,
                userId: decoded.payload.sub,
            })
        }
    } catch (err) {
        console.error(err)
        return generatePolicy({
            effect: 'Deny',
            resource: event.methodArn,
            userId: 'Unauthorized',
        })
    }
}

export default Authorize
