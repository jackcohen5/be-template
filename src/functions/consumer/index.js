import { deleteMessage } from 'services/SQS'

export const handler = async event => {
    for (const r of event.Records) {
        const {
            body: message,
            messageAttributes: {
                strParam: { stringValue: strParam },
            },
            receiptHandle,
        } = r

        try {
            console.log('Processing message...', message, strParam)
            await deleteMessage({ receiptHandle })
        } catch (e) {
            console.error('Handling error...', e)
        }
    }
}
