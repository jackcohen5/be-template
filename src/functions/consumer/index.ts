import { SQSHandler } from 'aws-lambda'

import { deleteMessage } from 'services/SQS'

export const handler: SQSHandler = async (event) => {
    for (const r of event.Records) {
        const {
            body: message,
            messageAttributes: {
                strParam: { stringValue: strParam },
            },
            receiptHandle,
        } = r

        try {
            console.log('Processing message...', message, strParam)
            await deleteMessage({ receiptHandle })
        } catch (e) {
            console.error('Handling error...', e)
        }
    }
}
