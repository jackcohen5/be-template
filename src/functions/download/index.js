import { Role2View } from 'functions/BaseView'
import { generateDownloadUrl } from 'services/S3'

const DownloadView = async ({
    auth: { userId },
    pathParameters: { fileId },
}) => {
    const key = `${userId}/${fileId}`
    const url = await generateDownloadUrl(key)
    return { data: url }
}

export default Role2View(DownloadView)
