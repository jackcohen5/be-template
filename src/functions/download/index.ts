import { FunctionHandler, Role2View } from 'functions/BaseView'
import { generateDownloadUrl } from 'services/S3'

interface DownloadPayload {
    url: string
}

const DownloadView: FunctionHandler<DownloadPayload> = async ({
    auth: { userId },
    pathParameters: { fileId },
}) => {
    const key = `${userId}/${fileId}`
    const url = await generateDownloadUrl(key)
    return { data: { url } }
}

export default Role2View(DownloadView)
