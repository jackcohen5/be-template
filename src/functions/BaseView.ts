import type {
    APIGatewayProxyEvent,
    APIGatewayProxyEventPathParameters,
    APIGatewayProxyEventQueryStringParameters,
    APIGatewayProxyResult,
} from 'aws-lambda'

import { errorResponse, successResponse } from 'services/Lambda'
import { Role } from 'functions/authorizer/constants'

const extractRoleFromEvent = (
    event: APIGatewayProxyEvent,
): Role | undefined => {
    return event.requestContext.authorizer?.jwt?.claims?.claims?.role
}

interface EnhancedEvent<RequestBody> {
    auth: Record<string, unknown>
    body: RequestBody
    pathParameters: APIGatewayProxyEventPathParameters
    queryParameters: APIGatewayProxyEventQueryStringParameters
}

interface HandlerResponse<ResponsePayload> {
    data: ResponsePayload
    statusCode?: number
}

export type FunctionHandler<
    ResponsePayload,
    RequestBody = Record<string, unknown>,
> = (
    event: EnhancedEvent<RequestBody>,
) => Promise<HandlerResponse<ResponsePayload>>

type View = (
    event: APIGatewayProxyEvent,
    authorizedRoles?: Role[],
) => Promise<APIGatewayProxyResult>

const ViewWrapper = async <RequestBody, ResponsePayload>(
    f: FunctionHandler<ResponsePayload, RequestBody>,
    event: APIGatewayProxyEvent,
): Promise<APIGatewayProxyResult> => {
    try {
        const enhancedEvent: EnhancedEvent<RequestBody> = {
            auth: event?.requestContext?.authorizer ?? {},
            body: JSON.parse(event?.body ?? '{}'),
            pathParameters: event?.pathParameters ?? {},
            queryParameters: event?.queryStringParameters ?? {},
        }
        const { data, statusCode = 200 } = await f(enhancedEvent)
        return successResponse(data, statusCode)
    } catch (err) {
        console.error(err)
        return errorResponse({ message: 'Unexpected error occurred.' })
    }
}

export const UnauthenticatedView =
    <RequestBody, ResponsePayload>(
        f: FunctionHandler<RequestBody, ResponsePayload>,
    ): View =>
    (event: APIGatewayProxyEvent) =>
        ViewWrapper(f, event)

export const AnonymousView =
    <RequestBody, ResponsePayload>(
        f: FunctionHandler<RequestBody, ResponsePayload>,
    ): View =>
    async (event: APIGatewayProxyEvent) => {
        if (extractRoleFromEvent(event)) {
            return errorResponse({ message: 'Unauthorized' }, 403)
        }

        return await ViewWrapper(f, event)
    }

const AuthenticatedView =
    <RequestBody, ResponsePayload>(
        f: FunctionHandler<RequestBody, ResponsePayload>,
        authorizedRoles: Role[] = [],
    ): View =>
    async (event: APIGatewayProxyEvent) => {
        if (authorizedRoles.includes(extractRoleFromEvent(event))) {
            return await ViewWrapper(f, event)
        } else {
            return errorResponse({ message: 'Unauthorized' }, 403)
        }
    }

export const Role1View = <RequestBody, ResponsePayload>(
    f: FunctionHandler<RequestBody, ResponsePayload>,
): View => AuthenticatedView(f, [Role.ROLE_1])

export const Role2View = <RequestBody, ResponsePayload>(
    f: FunctionHandler<RequestBody, ResponsePayload>,
): View => AuthenticatedView(f, [Role.ROLE_2])
