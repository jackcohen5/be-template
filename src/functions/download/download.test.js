import { Roles } from 'functions/BaseView'

import { authenticatedEvent, makeAuthenticatedEvent } from '../../../mocks'

import download from '.'

describe('download', () => {
    it('returns a 403 if role 1', async () => {
        const response = await download(authenticatedEvent)
        expect(response).toMatchSnapshot()
    })

    it('executes as expected if role 2', async () => {
        const response = await download(
            makeAuthenticatedEvent({ roles: [Roles.TEMPLATE_NAME_ROLE2] }),
        )
        expect(response).toMatchSnapshot()
    })
})
