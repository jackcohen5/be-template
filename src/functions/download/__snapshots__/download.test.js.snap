// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`download executes as expected if role 2 1`] = `
Object {
  "body": "{\\"data\\":\\"download-url\\"}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;

exports[`download returns a 403 if role 1 1`] = `
Object {
  "body": "{\\"message\\":\\"Unauthorized\\"}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 403,
}
`;
